discussion/ownership-and-mutable-aliases.tex

```diff
@@ -1,6 +1,103 @@
+
+This can be demonstrated with a new example.
+Let us say that we have a program in which we have configured two interrupts, \texttt{I1} and \texttt{I2}, to be triggered at differently timed intervals in the program.
+In the interrupt handler for \texttt{I1} we initiate a data transfer of \texttt{0x1111} over the \gls{usart}, and we initiate a data transfer of \texttt{0xFFFF} in the interrupt handler for \texttt{I2}.
+Now, what happens if the \texttt{I2} interrupt occur while the interrupt handler for \texttt{I1} is executing?
+If \texttt{I2} has higher priority than \texttt{I1}, the execution will change scope to \texttt{I2}'s interrupt handler, and a new \gls{usart} transfer will be initiated.
+Effectively, a race condition have occurred where two different interrupt handlers have acquired a handle to the same peripheral, and both are using this peripheral to transfer data at the same time.
+This might result in corrupted data being sent over the \gls{usart}.
+Even worse, the application at the receiving end of the \gls{usart} might end up with a program crash, if it relies on the data to be delivered in a certain format.
+
+\subsection{Inspiration?}
+% After the introduction to {\rust}'s ownership-rules presented in \autoref{sec:guaranteed_memory_safety}
+The question that we are left with after the previous section is then; is there a way to apply {\rust}'s ownership-semantics directly to the hardware, and let the {\rust} compiler do the reasoning for us of whether it is safe to use a peripheral or not?
+As it turns out, there is no easy or straight forward way to do this, at least not by our finding.
+
+One thing that can be attempted is to hide the hardware initialization process from the programmer and let it be done automatically by the compiler, similar to what {\zinc} does, as described in \autoref{sec:zinc}.
+{\zinc} uses the Platform Tree to specify the hardware which do the process of initializing the peripherals and save them into a \code{run\_args} structure which acts as their owner.
+This approach of initialization has two advantages:
+
+\begin{enumerate}[\hspace{13}1)]
+    \item Compile-time verification can guarantee that the peripherals gets initialized correctly.
+    \item The peripherals are owned by a variable, and access to these peripherals can be controlled by {\rust}'s borrow-checker.
+\end{enumerate}
+
+Point 1) is indeed a very interesting feature of {\zinc}, but it does not really help us to solve the problem that we are discussing in this section.
+Point 2) can help us to make sure that all peripheral access goes through the \code{run\_args} structure, but this structure is only passed to the program's main loop.
+What happens if we want to use a peripheral in an interrupt handler, which defines a different scope that is outside of the main loop, like in the example from the previous section?
```

Comment:
Refere to the irq section, and say that this is one way of handling the issue with Rusts ownership, but we should discuss limitation with this method as well.